File: leetcode/1.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

bool comparer(pair<int, int> i, pair<int, int> j)
{
    return (i.first < j.first);
}

// solution: sort nums and record the num's indices, then search from left and right
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<pair<int, int>> nums_indices;
        for (int i = 0; i < nums.size(); i++)
        {
            nums_indices.push_back(pair<int, int>(nums[i], i));
        }

        sort(nums_indices.begin(), nums_indices.end(), comparer);

        int left = 0;
        int right = nums.size() - 1;
        vector<int> res;

        while (left < right)
        {
            int sum = nums_indices[left].first + nums_indices[right].first;
            if (sum == target)
            {
                res.push_back(nums_indices[left].second);
                res.push_back(nums_indices[right].second);

                return res;
            }

            if (sum < target) left++;
            else right--;
        }

        return res;
    }
};

// using hash table
class Solution2 {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> indices;
        vector<int> res;

        for (int i = 0; i < nums.size(); i++)
        {
            if (indices.find(nums[i]) != indices.end() && nums[i] * 2 == target)
            {
                res.push_back(indices.find(nums[i])->second);
                res.push_back(i);

                return res;
            }

            indices[nums[i]] = i;
        }

        for (int i = 0; i < nums.size(); i++)
        {
            int n1 = nums[i];
            int n2 = target - n1;

            if (indices.find(n2) != indices.end() && n1 != n2)
            {
                res.push_back(indices[n1]);
                res.push_back(indices[n2]);

                return res;
            }
        }

        return res;
    }
};

int main()
{
    vector<int> nums = { 32,71,12,45,26,80,53,33 };

    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << endl;
    }

    vector<pair<int, int>> nums_indices;
    for (int i = 0; i < nums.size(); i++)
    {
        nums_indices.push_back(pair<int, int>(nums[i], i));
    }

    sort(nums_indices.begin(), nums_indices.end(), comparer);

    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << endl;
    }

    return 0;
}

File: leetcode/1.py
'''
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
UPDATE (2016/2/13):
The return format had been changed to zero-based indices. Please read the above updated description carefully.
'''

'''
Solution 1: sort the nums and record the original indices
'''


class Solution(object):
    """
    sort the nums, and record the original indices
    then search from left and right

    O(nlogn)
    """

    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        nums = [(nums[i], i) for i in range(len(nums))]
        nums = sorted(nums, key=lambda x: x[0])

        left, right = 0, len(nums) - 1
        while left < right:
            temp = nums[left][0] + nums[right][0]
            if temp == target:
                return [nums[left][1], nums[right][1]]
            if temp < target:
                left += 1
            else:
                right -= 1


'''
solution 2: by hash table
'''
class Solution(object):
    """
    O(n)
    using hash table to lookup another number in this hash table given another number,
    if another number is in hash table, then we could get number indices.
    """
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        indices = {}
        for i in xrange(len(nums)):
            if nums[i] in indices and nums[i] * 2 == target:
                return [indices[nums[i]], i]

            indices[nums[i]] = i

        for i in xrange(len(nums)):
            n1 = nums[i]
            n2 = target - n1
            if n2 in indices and n1 != n2:
                return [indices[n1], indices[n2]]
